djangocms_picture/models.py

```diff
@@ -1,83 +1,284 @@
+    use_no_cropping = models.BooleanField(
+        verbose_name=_('Use original image.'),
+        blank=True,
+        default=False,
+        help_text=_('Outputs the raw image without cropping.'),
+    )
+    # upscale and crop work together
+    # throws validation error if other cropping options are selected
+    use_crop = models.BooleanField(
+        verbose_name=_('Crop image'),
+        blank=True,
+        default=False,
+        help_text=_('Crops the image according to the given thumbnail settings in the template.'),
+    )
+    use_upscale = models.BooleanField(
+        verbose_name=_('Upscale image'),
+        blank=True,
+        default=False,
+        help_text=_('Upscales the image to the size of the thumbnail settings in the template.')
+    )
+    # overrides all other options
+    # throws validation error if other cropping options are selected
+    use_thumbnail = models.ForeignKey(
+        ThumbnailOption,
+        verbose_name=_('Thumbnail options'),
+        blank=True,
+        null=True,
+        help_text=_('Overrides width, height, crop and upscale with the provided preset.'),
+    )
+
+    # Add an app namespace to related_name to avoid field name clashes
+    # with any other plugins that have a field with the same name as the
+    # lowercase of the class name of this model.
+    # https://github.com/divio/django-cms/issues/5030
+    cmsplugin_ptr = models.OneToOneField(
+        CMSPlugin,
+        related_name='%(app_label)s_%(class)s',
+        parent_link=True,
+    )
 
     def __str__(self):
-        if self.alt:
-            return self.alt[:40]
-        elif self.image:
-            # added if, because it raised attribute error when file wasn't
-            # defined.
-            try:
-                return u"%s" % os.path.basename(self.image.name)
-            except AttributeError:
-                pass
-        return u"<empty>"
+        if self.picture and self.picture.label:
+            return self.picture.label
+        return str(self.pk)
+
+    def get_short_description(self):
+        if self.picture and self.picture.label:
+            return self.picture.label
+        if self.link_url:
+            return self.link_url
+        return ugettext('<file is missing>')
+
+    def copy_relations(self, oldinstance):
+        # Because we have a ForeignKey, it's required to copy over
+        # the reference from the instance to the new plugin.
+        self.picture = oldinstance.picture
+
+    def get_size(self, context, placeholder):
+        crop = self.use_crop or False
+        upscale = self.use_upscale or False
+        # use field thumbnail settings
+        if self.use_thumbnail:
+            width = self.use_thumbnail.width
+            height = self.use_thumbnail.height
+            crop = self.use_thumbnail.crop
+            upscale = self.use_thumbnail.upscale
+        elif self.use_automatic_scaling:
+            width = context.get('width', None)
```

Comment:
design practice, if your function or method asks for a width / height, then set those as parameters (optional in this case), instead of asking for the container of such values (context). It's the same concept as a permission function asking for a user vs asking for the entire request object. Always ask for what you need only.